### PR TITLE
Fix deconstruction in await-foreach statements

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/42770

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Diagnostics/AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest.cs [hunk 1]
```diff
@@ -15,5 +15,128 @@ public abstract partial class AbstractCSharpDiagnosticProviderBasedUserDiagnosti
...
+    {
+        IAsyncEnumerator<T> GetAsyncEnumerator();
+    }
+
+    public interface IAsyncEnumerator<out T> : IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask<bool> MoveNextAsync();
+        T Current { get; }
+    }
+}";
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest.cs]: this was hte minimal version i coudl get where everything worked.

File: src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -172,6 +172,8 @@ private void AnalyzeNode(SyntaxNodeAnalysisContext context)
                 return false;
             }
 
+            var diagnostics = semanticModel.GetDiagnostics();
+
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UseDeconstruction/CSharpUseDeconstructionDiagnosticAnalyzer.cs]:
```suggestion
```

File: src/EditorFeatures/CSharpTest/Diagnostics/AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/CSharpTest/Diagnostics/AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest.cs]: Is there a copy of this somewhere in the compiler layer for their own testing we can share?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Diagnostics/AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest.cs]: No, compiler has a copy that is very differnet and depends on hteir own facad libs unfortunately.

**Pull request merged**